### ISSUE TITLE
Assign a component instance to an Entity

### ISSUE THREAD

**Issue by joecons97** [skypjack/entt#1004]
Hi all,

I'm building a serialisation system and I'm little stuck on what is described in the title.

```c++
//Create entity
entt::entity ent = Scene::GetActiveScene()->GetRegistry().create();
//Loop through each component node from YAML
for (auto component : entity) {
        //Get component name from YAML node
	std::string componentName = component.first.as<std::string>();
        //Get type id from component name loaded from YAML
	auto typeId = entt::hashed_string{ componentName.c_str() };
        //Get meta_type from typeId
	entt::meta_type type = entt::resolve(typeId);
        //Create instance of component based on meta_type
	auto instance = type.construct();
        //Load data from YAML node into component instance (per component basis)
	type.func(entt::hashed_string{ "Deserialise" }).invoke(instance, component.second);
	//TODO add instance of component to entity
}
``` 

Is this even possible?
I've seen registry.assign being used in other issues but it seems like that function has changed since then.
I'm using 3.11.1

Any help on this would be appreciated.

**Comment by skypjack** [comment]
Sorry for the dumb question but what are `entity` and `component` (`first` and `second`) exactly?
I mean, what are their types? I can't easily parse your example nor I get where you're stuck.

> Is this even possible?

Yeah, definitely. 🙂

**Comment by joecons97** [comment]
> Sorry for the dumb question but what are `entity` and `component` (`first` and `second`) exactly? I mean, what are their types? I can't easily parse your example nor I get where you're stuck.
> 
> > Is this even possible?
> 
> Yeah, definitely. 🙂

Sorry, I really should have remained those variables before posting!

I've updated it now to be more clear (hopefully)😄

**Comment by skypjack** [comment]
Well, you've a couple of options here.
The first that comes to my mind is to also add an emplace-function to your meta type and literally use the typed `emplace` under the hood.
Otherwise, you can use the type info of your component with the `storage` function of the registry, get an opaque storage and set the element using  `push`.
I think the former is closer to what you're doing already but it's up to you to decide what you like more ofc. 🙂

**Comment by skypjack** [comment]
> Check if the registry storage actually contains that type

Do you mean checking if a pool exists?

> register a typeid into the registry storage

Can I ask you to elaborate a little further? 😅 
I don't get exactly what you mean with _register a typeid into the registry storage_, sorry. 🙏

**Comment by skypjack** [comment]
If you know the type, you can use the `storage` function to do that as in `registry.storage<T>()`.
Ironically, yesterday I started working on making this possible even when the C++ type is not known (for example if you are copying entities and components between different registries in an opaque way).

**Comment by joecons97** [comment]
> If you know the type, you can use the `storage` function to do that as in `registry.storage<T>()`. Ironically, yesterday I started working on making this possible even when the C++ type is not known (for example if you are copying entities and components between different registries in an opaque way).

That's the trouble, I only know the type of the component in the sense that I have its id_type but not its actual written/declared type and so I can't really use those template functions.

An example function of the sort of thing I'd need is
```c++
//Get component type info from reflection system
entt::meta_type componentType = entt::resolve(componentTypeId);
//Get type id
auto typeIdHash = componentType.info().hash();
//Get or create new pool based upon the component type id hash
auto componentPool = registry.storage().get_or_emplace_pool(typeIdHash); //<---- THIS
//Create entity & instance of component
auto ent = registry.create();
auto instance = componentType.construct();
//Add instance of component to entity
componentPool.second.emplace(ent, instance.data());
```

**Comment by skypjack** [comment]
Yeah, I see. I plan to release a sort of _fake vtable_ for opaque storage types in the registry with the next release. In the meantime, since you're using meta already, you can attach the storage function to the meta type for the component and use it when the storage isn't available. Have you tried it already?

**Comment by joecons97** [comment]
> Yeah, I see. I plan to release a sort of _fake vtable_ for opaque storage types in the registry with the next release. In the meantime, since you're using meta already, you can attach the storage function to the meta type for the component and use it when the storage isn't available. Have you tried it already?

I have not, how would I do that sorry?

**Comment by skypjack** [comment]
I'm a little on a rush and can't pack a working example on the fly, sorry. Anyway, when you bind the type, you've to register the function as in:
```cpp
.func<&entt::registry::storage<T>, entt::as_ref_t>("storage"_hs)
```
Probably with a `static_cast` to disambiguate the const vs non-const overload. When you've it, you can invoke the function as:
```cpp
auto &storage = entt::resolve<T>().invoke("storage"_hs, {} /* no component instance here, it's a static meta function */, entt::forward_as_meta(registry), id).cast<typename entt::registry::common_type &>();
```
Let me know if you've problems and I can prepare a working example during the next days.

**Comment by joecons97** [comment]
Sorry, have been super so only just got around to this!
I'm having issues with the func definition - am I correct in thinking that func must be defined for each component? or have I got that total wrong? 

If my thinking is correct then it's giving me the below error :(
```
C++ no instance of function template matches the argument list            argument types are: (entt::hashed_string)            object type is: entt::meta_factory<TransformComponent, std::integral_constant<Entity TransformComponent::*, &TransformComponent::parent>, std::integral_constant<Entity TransformComponent::*, &TransformComponent::parent>>
``` 

Here's how I'm defining it for my TransformComponent
```c++
entt::meta<TransformComponent>().type(entt::hashed_string{ "TransformComponent" }).base<Serialisable>()
	.ctor()
	.data<&TransformComponent::pos>(entt::hashed_string{ "pos" })
	.data<&TransformComponent::euler>(entt::hashed_string{ "euler" })
	.data<&TransformComponent::scale>(entt::hashed_string{ "scale" })
	.data<&TransformComponent::parent>(entt::hashed_string{ "parent" })
	.func<&entt::registry::storage<TransformComponent>, entt::as_ref_t> (entt::hashed_string{ "storage" });
```

**Comment by skypjack** [comment]
Ohooo my fault, completely missed the notification 🙏 I apologize.
Yeah, the function has to be defined for all component types. Definitely.
Do you have a small repro to reproduce exactly your error? I guess it's due to the fact that `storage` is overloaded. The fix can be as simple as a `static_cast` to help the compiler to disambiguate.

**Comment by joecons97** [comment]
No worries! :)
I tried the static_cast suggestion with no luck - admittedly I'm more of a C# guy - C++ isn't really my strong suit so I'm sure I'm just doing something stupid lol.

Here's a small repo
https://github.com/joecons97/entt-storage-repro

**Comment by skypjack** [comment]
Not a problem. [Here](https://godbolt.org/z/PPPv54qEP) we go. 👍

**Comment by joecons97** [comment]
Oh brilliant thank you so much! I was fairly close lol!

Now I can't get the storage function to invoke properly - I've been debugging it without the cast (so like this).

```c++
auto storageFunc = type.func(entt::hashed_string{ "storage" });
storageFunc.invoke({}, entt::forward_as_meta(registry), typeId);
``` 
and I've found that the function isn't even called! 

Stepping through the code, I can see that none of the conditions in ```internal::meta_invoke``` are met and thus it's just returning ```meta_any{meta_ctx_arg, ctx}```.

I'm starting to think I'm cursed with this haha

**Comment by skypjack** [comment]
What type is `typeId`? I'm doing exactly the same [here](https://godbolt.org/z/s6aE7eh45) and it works like a charm.

**Comment by joecons97** [comment]
> What type is `typeId`? I'm doing exactly the same [here](https://godbolt.org/z/s6aE7eh45) and it works like a charm.

Oooh I'm dumb lol - I had it stored as a hashed_string and just assumed it would auto-cast it! My bad! This is all working perfectly now, thank you so much for your help <3

**Comment by skypjack** [comment]
I see. No, well, when you pass something as a meta function parameter, the system wraps it in a `meta_any` and that's it. 👍
